src/main/java/com/google/devtools/build/lib/worker/WorkerSpawnRunner.java

```diff
@@ -523,17 +531,190 @@ WorkResponse execInWorker(
     return response;
   }
 
+  /**
+   * Initializes metricsSet for workers. If worker metrics already exists for this worker, does
+   * nothing
+   */
+  private void initializeMetricsSet(WorkerKey workerKey, Worker worker) {
+
+    if (workerIdToWorkerMetric.containsKey(worker.getWorkerId())) {
+      return;
+    }
+    long processId = worker.getProcessId();
+
+    WorkerMetric workerMetric =
+        new WorkerMetric(
+            worker.getWorkerId(),
+            processId,
+            workerKey.getMnemonic(),
+            workerKey.isMultiplex(),
+            workerKey.isSandboxed());
+
+    workerIdToWorkerMetric.put(worker.getWorkerId(), workerMetric);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **540 and 553**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).